src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,159 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
+
+public class MyLinkedList<E> implements MyLinkedListInterface<E> {
+    private Node<E> head;
+    private Node<E> tail;
+    private int size;
 
-public class MyLinkedList<T> implements MyLinkedListInterface<T> {
     @Override
-    public void add(T value) {
+    public void add(E value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            ++size;
+            return;
+        }
+        Node<E> newNode = new Node<>(tail, value, null);
+        tail.next = newNode;
+        tail = newNode;
+        ++size;
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(E value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<E> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            ++size;
+            return;
+        }
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        Node<E> add = new Node<>(newNode.prev, value, newNode);
+        newNode.prev.next = add;
+        newNode.prev = add;
+        size++;
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public void addAll(List<E> list) {
+        for (E element : list) {
+            add(element);
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public E get(int index) {
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        return newNode.value;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public E set(E value, int index) {
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        E result = newNode.value;
+        newNode.value = value;
+        return result;
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public E remove(int index) {
+        checkIndex(index);
+        Node<E> newNode = findNode(index);
+        E removeValue = newNode.value;
+        unlink(newNode);
+        return removeValue;
```

Comment:
```suggestion
        return unlink(findNode(index));
```
якщо ми повертаємо велью з анлінку можна спростити